### PR TITLE
Add 'edge' and 'reflect' mode to cupy.pad

### PR DESCRIPTION
**Pull request by wkentaro**
## Why?

Currently, `cupy.pad` only accepts `mode='constant'` and I'd like to use it with `mode='reflect'`.

## Concern

Mostly, I copied some functions from `numpy.pad` and should I write the notice for it?
Or the declaration in `LICENSE` file is enough?

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -2350,7 +2350,7 @@ cpdef ndarray _repeat(ndarray a, repeats, axis=None):
     # Scalar and size 1 'repeat' arrays broadcast to any shape, for all
     # other inputs the dimension must match exactly.
     cdef bint broadcast = False
-    if isinstance(repeats, int):
+    if numpy.issubdtype(repeats, numpy.integer):
```
Comment (wkentaro) [inline cupy/core/core.pyx]: This was necessary because `cupy.pad` internally uses `cupy.repeat` and
the `repeats` argument can have numpy int types.

**Comment by wkentaro** [comment]
Thank you! I updated the code.

**Comment by okuta** [comment]
Thanks.

jenkins, test this please.

**Comment by okuta** [comment]
I checked your PR.
I got errors on some environment.
For example, `numpy==1.13.3`. The error log is [here](https://gist.github.com/okuta/3eeb93c8296493c3af1d533e9a812b07).

Please check  with old NumPy (1.9, 1.10, 1.11, 1.12, 1.13 ).

**Comment by okuta** [comment]
Thanks.
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Comment by wkentaro** [comment]
Thanks!

**Pull request merged**